cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/servlet/UserTaskManager.java

```diff
@@ -0,0 +1,456 @@
+      }
+    } else {
+      if (httpServletRequest.getHeader(USER_TASK_HEADER_NAME) != null) {
+        // request provides user_task_header that user tasks doesn't exist then just throw exception
+        String userTaskIdFromRequest = httpServletRequest.getHeader(USER_TASK_HEADER_NAME);
+        throw new IllegalArgumentException(
+            String.format("UserTask %s is an invalid %s", userTaskIdFromRequest, USER_TASK_HEADER_NAME));
+      }
+
+      if (step != 0) {
+        throw new IllegalArgumentException(
+            String.format("There are no step in the session. Cannot add step %d.", step));
+      }
+
+      SessionKey sessionKey = new SessionKey(httpServletRequest);
+      userTaskId = _uuidGenerator.randomUUID();
+      LOG.info("Create a new UserTask {} with SessionKey {}", userTaskId, sessionKey);
+
+      OperationFuture future = operation.get();
+      insertFuturesByUserTaskId(userTaskId, future, httpServletRequest);
+
+      synchronized (_sessionToUserTaskIdMap) {
+        _sessionToUserTaskIdMap.put(sessionKey, userTaskId);
+      }
+
+      httpServletResponse.setHeader(USER_TASK_HEADER_NAME, userTaskId.toString());
+      return future;
+    }
+  }
+
+  @SuppressWarnings("unchecked")
+  public <T> T getFuture(HttpServletRequest request) {
+    UUID userTaskId = getUserTaskId(request);
+    List<OperationFuture> operationFutures = getFuturesByUserTaskId(userTaskId, request);
+    if (operationFutures == null || operationFutures.isEmpty()) {
+      return null;
+    }
+
+    return (T) operationFutures.get(operationFutures.size() - 1);
+  }
+
+  public void closeSession(HttpServletRequest request) {
+    SessionKey sessionKey = new SessionKey(request);
+    UUID userTaskId;
+    synchronized (_sessionToUserTaskIdMap) {
+      userTaskId = _sessionToUserTaskIdMap.remove(sessionKey);
+    }
+
+    if (userTaskId != null) {
+      LOG.info("Closing SessionKey {} and UserTaskId {}", sessionKey, userTaskId);
+      if (isActiveUserTasksDone(userTaskId)) {
+        LOG.info("Invalidate SessionKey {}", sessionKey);
+        sessionKey.httpSession().invalidate();
+      }
+    }
+  }
+
+  private void expireOldSessions() {
+    long now = _time.milliseconds();
+    synchronized (_sessionToUserTaskIdMap) {
+      Iterator<Map.Entry<SessionKey, UUID>> iter = _sessionToUserTaskIdMap.entrySet().iterator();
+      while (iter.hasNext()) {
+        Map.Entry<SessionKey, UUID> entry = iter.next();
+        SessionKey sessionKey = entry.getKey();
+        HttpSession session = sessionKey.httpSession();
+        LOG.trace("Session {} was last accessed at {}, age is {} ms", session, session.getLastAccessedTime(),
+            now - session.getLastAccessedTime());
+        if (now >= session.getLastAccessedTime() + _sessionExpiryMs) {
+          LOG.info("Expiring SessionKey {}", entry.getKey());
+          iter.remove();
+          session.invalidate();
+          // NOTE: Does it make sense to cancel the future?
```

Comment:
Please feel free to remove the comment if you prefer. Based on our offline discussion, I agree that it is better not to cancel the future.